### PR TITLE
Deprecate lenient parsing of bulk actions

### LINKED ISSUE #43774: Bulk REST API allows arbitrary fields after action
While trying to refresh a specific index within a `bulk` request containing multiple indices, I came across an unexpected scenario where unrecognized field names were ignored:

```http
POST /test/_bulk
{"index":{"_id":"1"}, "abc": true}
{"field":"value"}
```

In ES 7.2.0, at least, this results in the document being indexed.

However, if you put `abc` in front of the `index` action, then it triggers an expected error [from my perspective]:

```http
POST /test/_bulk
{"abc": true, "index":{"_id":"1"}}
{"field":"value"}
```

results in

```json
{
  "error": {
    "root_cause": [
      {
        "type": "illegal_argument_exception",
        "reason": "Malformed action/metadata line [1], expected START_OBJECT or END_OBJECT but found [VALUE_BOOLEAN]"
      }
    ],
    "type": "illegal_argument_exception",
    "reason": "Malformed action/metadata line [1], expected START_OBJECT or END_OBJECT but found [VALUE_BOOLEAN]"
  },
  "status": 400
}
```

We just need to update the parser to ensure that the JSON object is empty after finding the action. I can imagine that some people think some follow-on fields are taking effect as it is today.

### PR DESCRIPTION
**Pull request by arteam**
Make sure there are no arbitrary fields after an action declaration and that it gets properly closed by a curly bracket.

Resolves #43774

### CODE REVIEW

**Review by DaveCTurner — CHANGES_REQUESTED** [review]
I suggested some other ways that we accept invalid input. Could you add tests that assert that the new warnings are emitted too please?

File: server/src/main/java/org/elasticsearch/action/bulk/BulkRequestParser.java [hunk 1]
```diff
@@ -406,6 +424,35 @@ public void parse(
...
+                STRICT_ACTION_PARSING_WARNING_KEY,
+                "A bulk action object contained multiple keys. Additional keys are currently ignored but will be rejected in a "
+                    + "future version."
+            );
+            return;
+        }
+        if (parser.nextToken() != null) {
+            deprecationLogger.compatibleCritical(
+                STRICT_ACTION_PARSING_WARNING_KEY,
+                "A bulk action contains trailing junk after the closing brace. It is currently ignored but will be rejected in a "
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/action/bulk/BulkRequestParser.java:450]: Rewording suggestion:

```suggestion
                "A bulk action contained trailing data after the closing brace. This is currently ignored but will be rejected in a "
```

File: server/src/main/java/org/elasticsearch/action/bulk/BulkRequestParser.java [hunk 2]
```diff
@@ -179,6 +185,17 @@ public void parse(
...
+                if (SUPPORTED_ACTIONS.contains(action) == false) {
+                    deprecationLogger.compatibleCritical(
+                        STRICT_ACTION_PARSING_WARNING_KEY,
+                        String.format(
+                            Locale.ROOT,
+                            "Unsupported action: `%s`. Supported values are `create`, `delete`, `index`, and `update`. "
+                                + "Unsupported actions are currently accepted but will be rejected in a future version.",
+                            action
+                        )
+                    );
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/action/bulk/BulkRequestParser.java:197]: `DeprecationLogger#compatibleCritical` looks to accept a format string and args already, do we need to use `String.format` here or could we just do this?

```suggestion
                    deprecationLogger.compatibleCritical(
                        STRICT_ACTION_PARSING_WARNING_KEY,
                            "Unsupported action: [{}]. Supported values are [create], [delete], [index], and [update]. "
                                + "Unsupported actions are currently accepted but will be rejected in a future version.",
                            action
                    );
```

Also I'd prefer `[brackets]` over backticks in these messages.

**Review by DaveCTurner — COMMENTED** [review]
Looks great, I left only minor comments/questions.

**Comment by arteam** [comment]
@elasticmachine update branch

**Review by DaveCTurner — COMMENTED** [review]
Code changes all LGTM, but since this adds deprecations I think we need to add docs about the change too. https://github.com/elastic/elasticsearch/pull/80410 adds the deprecation docs files for 8.1.0 so you can either wait for that to be merged or else duplicate those files here (and I'll sort out any merge conflicts on my PR later).

**Comment by arteam** [comment]
@elasticmachine update branch

**Comment by arteam** [comment]
Thanks David!

**Comment by DaveCTurner** [comment]
I would have preferred not to have merged this until the docs changes were reviewed and approved - it's hard to keep track of outstanding things to review otherwise. I opened https://github.com/elastic/elasticsearch/pull/80580 to suggest alternative wording and revert the unnecessary docs formatting changes that this PR introduces.

**Pull request merged**